code/week7/main.cpp

```diff
@@ -0,0 +1,186 @@
+    queue<char> queue;
+    queue.push('r');
+    queue.push('o');
+    queue.push('b');
+    queue.push('o');
+    queue.push('j');
+
+    while(!queue.empty()) {
+        cout << queue.front();
+        queue.pop();
+    }
+
+    /*************** STACK ***************/
+    // http://www.cplusplus.com/reference/stack/stack/
+    stack<char> stack;
+    stack.push('s');
+    stack.push('t');
+    stack.push('e');
+    stack.push('k');
+    stack.push('c');
+    stack.push('a');
+
+    while(!stack.empty()) {
+        cout << stack.top();
+        stack.pop();
+    }
+    cout << "\n\n";
+
+    /*************** SETS ***************/
+    // http://www.cplusplus.com/reference/set/set/?kw=set
+    // init
+    // set<TYPE> name
+    set<string> ordered_set {"1", "5", "6"};
+
+    // can insert elements into set
+    ordered_set.insert("2");
+    print_data(ordered_set.begin(), ordered_set.end(), "ordered_set");
+
+    // If the same element is inserted TODO
+    ordered_set.insert("1");
+    print_data(ordered_set.begin(), ordered_set.end(), "ordered_set");
+
+    // elements can also be erased
+    ordered_set.erase("1");
+    print_data(ordered_set.begin(), ordered_set.end(), "ordered_set");
+
+    /*************** UNORDERED SETS ***************/
+    // http://www.cplusplus.com/reference/unordered_set/unordered_set/
+    // init
+    // set<TYPE> name
+    unordered_set<string> unordered_set {"20","19","18","1","16"};
+
+    // can insert elements into set
+    unordered_set.insert("21");
+    // order is not maintained here
+    print_data(unordered_set.begin(), unordered_set.end(), "unordered_set");
+
+    /*************** MAPS ***************/
+    // http://www.cplusplus.com/reference/map/map/?kw=map
+    // uses set in background
+    cout << "\nmap\n";
+    // init
+    // map<KEY_TYPE, VALUE_TYPE>
+    map<char, int> char_map;
+
+    // inserts a pair (key, value) into the map at iterator
+    char_map.insert(char_map.begin(), pair<char, int>('a', 5));
+
+    // at gets the value at the given key
+    cout << "at char 'a' value = " << char_map.at('a') << endl;
+
+    /*************** UNORDERED MAPS ***************/
+    // http://www.cplusplus.com/reference/unordered_map/unordered_map/
+    // unordered maps have faster access time than map
+    // uses unordered set in background
+    unordered_map<int, string> string_map;
+
+    string_map.insert(pair<int, string>(53, "string here"));
+
+    cout << "at int 53 value = " << string_map.at(53) << endl;
+
+    /*************** TEMPLATES ***************/
+    cout << "\n\n";
+    cout << "the max of 7,8 is = " << getMax<int>(7, 8) << endl;
+    cout << "the max of 7.0, 8.0 is = " << getMax<double>(7.0, 8.0) << endl;
+    cout << "the min of 7, 8.0 is = " << getMin<int, double>(7, 8.0) << endl;
+
+    unique_ptr<Example<string>> example_str1 = make_unique<Example<string>>("storing strings here");
+    unique_ptr<Example<int>> exmaple2 = make_unique<Example<int>>(2);
```

Comment:
You never actually read the data from this level of the program, maybe adding that is a good idea.